### PR TITLE
Allow overriding `get_edit_handler` and `get_form_class` via `SnippetViewSet`

### PR DESCRIPTION
**Pull request by laymonage**
Part of wagtail/rfcs#85. Incorporates #10290 and #10294.

_Please check the following:_

-   [x] Do the tests still pass?[^1]
-   [x] Does the code comply with the style guide?
    -   [x] Run `make lint` from the Wagtail root.
-   [x] For Python changes: Have you added tests to cover the new/fixed behaviour?
-   [x] For new features: Has the documentation been updated accordingly?

**Please describe additional details for testing this change**.

Using bakerydemo, move the `panels` definition on the `Person` model to a custom `PersonViewSet`. Continuing the test example in #10276:

```py
from wagtail.snippets.models import register_snippet
from wagtail.snippets.views.snippets import SnippetViewSet


class PersonViewSet(SnippetViewSet):
    icon = "user"
    list_per_page = 2
    chooser_per_page = 1
    admin_url_namespace = "some_namespace"
    base_url_path = "deep/within/the/admin"
    chooser_admin_url_namespace = "my_chooser_namespace"
    chooser_base_url_path = "choose/wisely"
    # this will generate exact filters
    list_filter = ["job_title", "first_name"]
    # alternatively, this will generate filters with lookups
    list_filter = {
        "job_title": ["contains"],
        "first_name": ["contains"],
        "first_published_at": ["date"],
    }
    ordering = "job_title"
    # alternatively
    ordering = ["job_title", "last_name", "first_name"]

    def get_queryset(self, request):
        return self.model._default_manager.all().exclude(job_title__contains="[HIDDEN]")

    panels = [
        MultiFieldPanel(
            [
                FieldRowPanel(
                    [
                        FieldPanel("first_name"),
                        FieldPanel("last_name"),
                    ]
                )
            ],
            "Name",
        ),
        FieldPanel("job_title"),
        FieldPanel("image"),
        PublishingPanel(),
    ]


register_snippet(Person, viewset=PersonViewSet)
```

[^1]: [Development Testing](https://docs.wagtail.org/en/latest/contributing/developing.html#testing)
[^2]: [Browser and device support](https://docs.wagtail.org/en/latest/contributing/developing.html#browser-and-device-support)
[^3]: [Accessibility Target](https://docs.wagtail.org/en/latest/contributing/developing.html#accessibility-targets)

### CODE REVIEW

File: wagtail/admin/views/generic/mixins.py [hunk 1]
```diff
@@ -129,12 +129,14 @@ def get_context_data(self, **kwargs):
 
 
 class PanelMixin:
```
Comment (laymonage) [inline wagtail/admin/views/generic/mixins.py:131]: This should really be `EditHandlerMixin` if we want to keep the naming of "edit handler" for the top-level/root panel.

This mixin was made around the time the `EditHandler` class was renamed to `Panel`. I assumed we would completely drop the "edit handler" term, hence I named it `PanelMixin`.

File: wagtail/snippets/views/snippets.py [hunk 2]
```diff
@@ -868,6 +868,8 @@ def edit_view(self):
             template_name=self.get_edit_template(),
             header_icon=self.icon,
             permission_policy=self.permission_policy,
+            panel=self._edit_handler,
+            form_class=self.get_form_class(),
```
Comment (laymonage) [inline wagtail/snippets/views/snippets.py:872]: Not that it makes a difference, but I forgot to add `for_update=True` here (and in the `revisions_revert` view)

File: docs/topics/snippets.md [hunk 3]
```diff
@@ -32,15 +32,15 @@ class Advert(models.Model):
         return self.text
 ```
 
-The `Advert` model uses the basic Django model class and defines two properties: text and URL. The editing interface is very close to that provided for `Page`-derived models, with fields assigned in the `panels` property. Snippets do not use multiple tabs of fields, nor do they provide the "save as draft" or "submit for moderation" features.
+The `Advert` model uses the basic Django model class and defines two properties: text and URL. The editing interface is very close to that provided for `Page`-derived models, with fields assigned in the `panels` (or `edit_handler`) property. Snippets do not use multiple tabs of fields, nor do they provide the "save as draft" or "submit for moderation" features.
```
Comment (laymonage) [inline docs/topics/snippets.md]: The last line should've been updated to reflect the current state. I could do that now or as part of the docs update sometime during the feature freeze

Comment (laymonage) [reply docs/topics/snippets.md]: Oh never mind, I already did in https://github.com/wagtail/wagtail/commit/6cb13f9c6f727e1042da2e15c59d1f0ad93f401c

File: wagtail/admin/views/generic/mixins.py (hunk 1 above)
Comment (gasman) [reply wagtail/admin/views/generic/mixins.py:131]: Either name works for me really - it's the mixin that provides a top-level edit handler property, or the one that enables panel-based editing as a feature.

**Review by gasman — COMMENTED** [review]
LGTM!

**Review by gasman — APPROVED** [review]

**Comment by gasman** [comment]
Merged in 26a14da3d7cf743d13876a41bccb2e6c690f4de3 + parents

**Pull request closed without merging**